tools/flang1/flang1exe/semutil2.c

```diff
@@ -11916,6 +11916,16 @@ get_parm_ast(int parent, SPTR sptr, DTYPE dtype)
       return ast;
     }
   }
+  /* Field not found, so try again and take more recursive approach */
+  for (mem = DTY(dtype + 1); mem > NOSYM; mem = SYMLKG(mem)) {
+    /* Act only on the cases that were not considered before */
+    if ((!PARENTG(mem)) && strcmp(SYMNAME(sptr), SYMNAME(mem))) {
```

Comment:
Could the real issue be that the `PARENTG` field is not set for some cases?

Assuming PARENT is for finding the parent type in an inheritance/extends hierarchy, this might only be required at the leafs in which case the recursion is not needed.

Reply:
I don't think so. I've added this check for the `!PARENTG(mem)` just to prevent doing the recursive call to `get_param_ast()` for the second time. In this place we already know it won't find anything. Initially, there was no `!PARENTG(mem)` here and it worked the same, so I doubt `PARENTG` field is involved.

Reply:
Reposting a message shared internally here.

I was asking because the m5.f90 program work without this patch while m6.f90 needs this patch. m5.f90 and m6.f90 given below.

It will be helpful to proceed with the review if an explanation is provided for the above as well as a testcase demonstrating the change in the AST.
```
~/classic_flang$ cat m5.f90
program move_alloc_test

  implicit none
  type bar
    integer, allocatable :: x
  end type
  type some
    type(bar), allocatable :: a(:)
  end type
  type some_more
    type(some) :: b
  end type
  type some_more_more
    type(some_more) :: c
  end type
  type(bar), allocatable :: a(:)
  type(some_more_more) :: e
  allocate(a(128))
  call move_alloc(e%c%b%a,a)
end program
```
```
~/classic_flang$ cat m6.f90
program move_alloc_test

  implicit none
  type bar
    integer, allocatable :: x
  end type
  type some
    integer, allocatable :: a(:)
  end type
  type some_more
    type(some) :: b
  end type
  type some_more_more
    type(some_more) :: c
  end type
  integer, allocatable :: b(:)
  type(some_more_more) :: e
  allocate(b(128))
  call move_alloc(e%c%b%a, b)
end program
```

Reply:
@kiranchandramohan those two examples execute completely different execution paths, and I suspect the m5.f90 example exposes a different bug that is not relevant to this pull request. The code corrected by this patch is never reached for m5.f90 as the allocatable array of a derived type seems to be wrongly treated.

Let's start with a simple trace: the patch in this PR corrects the `get_parm_ast()` function. In case of m6.f90 the trace is following:

transform_move_alloc() calls copy_surrogate_to_bnds_vars() which calls add_parent_to_bounds() which eventually calls get_parm_ast().

In order for transform_move_alloc() to call copy_surrogate_to_bnds_vars(), the following condition must be held:
```
if (A_SHAPEG(fptr2) && sptr != sptr2 && !SDSCG(sptr2)) {
...
}
```
...where `sptr` represents the first param to `move_alloc()`, while `sptr2` represents the second param to `move_alloc()`.
`A_SHAPEG(fptr2)` is true in case of both m5.f90 and m6.f90, `sptr != sptr2` is true by nature, the only issue is with `SDSCG(sptr2)`. This macro checks if there is a static descriptor available for the second parameter. In case of the allocatable arrays, this should never be available and `!SDSCG(sptr2)` should yield true for both m5.f90 and m6.f90. Thing is, it doesn't. For some reason, static descriptor seems to be available for the allocatable array of a derived type! This simplifies the execution paths since the buggy function that this PR tries to fix is never being executed.